In [1]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

import warnings
warnings.filterwarnings(action='ignore')


In [2]:
# Read vehicles.csv
vehicles_data = pd.read_csv('./vehicles.csv')

In [3]:
# Read data
vehicles_data

Unnamed: 0     policy_id  car_no ownership_type  color  age  \
0                6  policy_74571       1         leased  other  4.0   
1                7  policy_74571       2          owned   blue  3.0   
2               15   policy_1998       1          owned   gray  7.0   
3               29  policy_28085       1          owned   gray  4.0   
4               34  policy_64282       1          owned  black  4.0   
...            ...           ...     ...            ...    ...  ...   
169232      345054  policy_34204       4          owned   gray  6.0   
169233      345055  policy_34204       5         loaned  black  8.0   
169234      345060    policy_621       1          owned  other  6.0   
169235      345061    policy_621       2          owned  black  3.0   
169236      345062    policy_621       3         loaned  other  4.0   

                     make_model  
0                 BMW : R1200CL  
1                    ACURA : TL  
2                   BMW : 750LI  
3       MERCEDES-BENZ : G55 AMG  
4                 BMW : 550I GT  
...                         ...  
169232            TOYOTA : RAV4  
169233        GMC : SIERRA 1500  
169234          DODGE : CHARGER  
169235        GMC : SAVANA 4500  
169236       HONDA : CB900F 919  

[169237 rows x 7 columns]

In [4]:
# delete the first column because there is no definition of the colum.

vehicles_data.drop(['Unnamed: 0'], axis='columns', inplace=True)

In [5]:
vehicles_data.head()

policy_id  car_no ownership_type  color  age               make_model
0  policy_74571       1         leased  other  4.0            BMW : R1200CL
1  policy_74571       2          owned   blue  3.0               ACURA : TL
2   policy_1998       1          owned   gray  7.0              BMW : 750LI
3  policy_28085       1          owned   gray  4.0  MERCEDES-BENZ : G55 AMG
4  policy_64282       1          owned  black  4.0            BMW : 550I GT

In [6]:
# check if there is any missing value
vehicles_data.isna().sum()

policy_id            0
car_no               0
ownership_type       0
color             1072
age                536
make_model           0
dtype: int64

As we can see there are 1072 missing color information and 536 missing age(of car) information.

In [7]:
# percentage of raw includes missing value out of total data.
vehicles_data[vehicles_data['color'].isna() | vehicles_data['age'].isna()].shape[0] / vehicles_data.shape[0]

0.009460106241542926

In [8]:
# data that includes missing value. (1072 samples)
vehicles_data[vehicles_data['color'].isna()]

policy_id  car_no ownership_type color  age  \
77      policy_77901       3          owned   NaN  4.0   
245     policy_37794       3          owned   NaN  7.0   
515     policy_60534       4          owned   NaN  2.0   
577     policy_36358       1         loaned   NaN  7.0   
637     policy_14342       2          owned   NaN  9.0   
...              ...     ...            ...   ...  ...   
168116  policy_29878       3         leased   NaN  7.0   
168174  policy_38226       2          owned   NaN  4.0   
168341  policy_30342       1          owned   NaN  4.0   
168342  policy_58543       1          owned   NaN  7.0   
169111  policy_74020       3          owned   NaN  4.0   

                        make_model  
77                      BMW : 535D  
245                     ACURA : RL  
515       MERCEDES-BENZ : ML63 AMG  
577          MERCEDES-BENZ : SL600  
637                     BMW : 135I  
...                            ...  
168116        FORD : GRAND MARQUIS  
168174  FORD : EXPLORER SPORT TRAC  
168341                 FORD : LOBO  
168342             HONDA : ODYSSEY  
169111              BUICK : ENCORE  

[1072 rows x 6 columns]

Percentage of raws including missing value is approximately 0.95% out of total.

How we can deal with those?

1. simply deleting raws including any missing value.
2. imputation using mean, median or mod.

For missing color information, we can observe what it the model of car because there is no any missing vlaue for the make_model column.

In [9]:
# number of samples do not have color information.
list_missing_color_cars = list(vehicles_data[vehicles_data['color'].isna()==True]['make_model'].unique())

581

In [9]:
# imputation using mod of color of each car type
dict_mod_colors = {}
for i in list_missing_color_cars:
    dict_mod_colors[i] = vehicles_data[vehicles_data['make_model']==i]['color'].value_counts().index.tolist()[0]
    
for i in range(len(vehicles_data)):
    
    if type(vehicles_data.loc[i]['color'])==float:
        
        vehicles_data.loc[i, ('color')]=dict_mod_colors[vehicles_data.loc[i]['make_model']]

NameError: name 'list_missing_color_cars' is not defined

We can see the each car has several different colors, so we can find mode of color each car and we can use that for missing car type.

For age, I think it is reasonable to use just mean of all age to fill out missing age.

In [ ]:
# now we don't have any missing color data
vehicles_data.isna().sum()

In [15]:
print(vehicles_data.loc[77])
print()
print(dict_mod_colors['BMW : 535D'])

policy_id         policy_77901
car_no                       3
ownership_type           owned
color                     blue
age                        4.0
make_model          BMW : 535D
Name: 77, dtype: object

blue


In [16]:
print(vehicles_data.loc[577])
print()
print(dict_mod_colors['MERCEDES-BENZ : SL600'])

policy_id                  policy_36358
car_no                                1
ownership_type                   loaned
color                              blue
age                                 7.0
make_model        MERCEDES-BENZ : SL600
Name: 577, dtype: object

blue


In [ ]:
# imputation using mean of age
vehicles_data = vehicles_data.fillna(vehicles_data.mean()['age'])

In [ ]:
# no missing value
vehicles_data.isna().sum()

In [19]:
vehicles_data

policy_id  car_no ownership_type  color  age  \
0       policy_74571       1         leased  other  4.0   
1       policy_74571       2          owned   blue  3.0   
2        policy_1998       1          owned   gray  7.0   
3       policy_28085       1          owned   gray  4.0   
4       policy_64282       1          owned  black  4.0   
...              ...     ...            ...    ...  ...   
169232  policy_34204       4          owned   gray  6.0   
169233  policy_34204       5         loaned  black  8.0   
169234    policy_621       1          owned  other  6.0   
169235    policy_621       2          owned  black  3.0   
169236    policy_621       3         loaned  other  4.0   

                     make_model  
0                 BMW : R1200CL  
1                    ACURA : TL  
2                   BMW : 750LI  
3       MERCEDES-BENZ : G55 AMG  
4                 BMW : 550I GT  
...                         ...  
169232            TOYOTA : RAV4  
169233        GMC : SIERRA 1500  
169234          DODGE : CHARGER  
169235        GMC : SAVANA 4500  
169236       HONDA : CB900F 919  

[169237 rows x 6 columns]

## Handling Outliers

In [ ]:
plt.rc('font', size=9)  
plt.subplots_adjust(right=1)


plt.subplot(221)
vehicles_data.ownership_type.value_counts(normalize=True).plot.pie()

plt.subplot(222)
vehicles_data.color.value_counts(normalize=True).plot.pie()

# plt.subplot(223)
# vehicles_data.car_no.value_counts(normalize=True).plot.barh()

plt.subplot(223)
vehicles_data.age.value_counts(normalize=True).plot.barh()

plt.show()

In [21]:
vehicles_data.car_no.describe()

count    169237.000000
mean          2.485231
std           1.345593
min           1.000000
25%           1.000000
50%           2.000000
75%           3.000000
max           8.000000
Name: car_no, dtype: float64

In [ ]:
vehicles_data.age.describe()

In [ ]:
sns.pairplot(data=vehicles_data, vars=['car_no', 'ownership_type', 'color', 'age'])

In [ ]:
sns.heatmap(vehicles_data[['age', 'car_no']].corr(), annot=True, cmap = 'Reds')
plt.show()

In [41]:
# for i in list(vehicles_data.groupby('policy_id')['car_no'].max()):
#     print(i, list(vehicles_data.groupby('policy_id')['car_no'].max()).count(i))

In [44]:
vehicles_data.groupby('policy_id')['car_no'].max()

policy_id
policy_100      6
policy_1000     2
policy_10001    2
policy_10002    5
policy_10004    2
               ..
policy_99994    4
policy_99996    4
policy_99997    4
policy_99998    1
policy_99999    6
Name: car_no, Length: 49162, dtype: int64

In [ ]:
# temp includes number of cards included in each policy
temp = list(vehicles_data.groupby('policy_id')['car_no'].max())

In [ ]:
# statistics of the number of cars in each policy.
print(pd.DataFrame(temp).describe())

plt.hist(temp)
plt.show()